### DISCUSSION TITLE
Hybrid Search Behaviour

### DISCUSSION THREAD

**Discussion by generatives** [milvus-io/milvus#12927] [Q&A and General discussion] [answered]
Hello!

I have a question about the behaviour of the hybrid search system. Does Milvus apply the filter while searching for the top K vectors (returns the the K top vectors which pass the filter) or does it apply the filter clause after finding the top K (returns the members of the top K vectors that pass the filter).

Thanks for any information.

**Comment by yhmo** [comment, marked as answer]
The Milvus filtering behavior is: 
First step: filtering vectors by the expression condition
Second step: do ANN search in the vectors which is filtered out, return topk items

**Reply by generatives** [reply]
That is very helpful, thank you!

**Reply by generatives** [reply]
Does that mean Milvus needs to do a pass through the entire candidate set with the filter before doing an ANN search? That sounds very expensive. Is there some kind of index or view we can use to speed that up?

**Reply by yhmo** [reply]
Yes, the filtering is expensive. Currently the performance is not good I think, but we will continue improve it.
Some details for this area:
- The milvus split data into segments, we call a row as "entity", a segment has many entities. An entity can contains vectors or scalar values, depends how many fields are defined to this collection. For example, we have a collection with two fields: [name: "id", type: int64], [name: "embedding", type: float_vector], then each entity will contain a int64 value and a vector.
- Each time the client send a search request, milvus will do ANN search for each segment, one by one. Each segment output a topk result, finally these topk results are combined into a final result and returned to client.
- The search engine has an ability to skip some vectors by a bitset, in fact the bitset is a bool array like [0, 1, 0, 0, 1, 1, 1, 0, ......]. The array size is equal to the segment entities count. If a bit is 0, the search engine will skip this vector.
- So, the filtering task is before the ANN search, to generate a bitset array so that the search engine can skip some vectors. The filtering task will travel each entity to see whether it can match the filter condition, then construct a bitset array. Now we haven't do much optimize for filtering task, it is more like a brute-force traverse. But yes we can do more optimize to speed up it. Possible approaches: 
    1) use meta data to quickly exclude
    2) use parallel computing to travel
    3) maybe hash can help
    4) string index(for string type field in v2.1)

**Reply by failable** [reply]
Is the details still true now? Can you point me to the related documentation? Thanks.

**Reply by yhmo** [reply]
Yes, the major logic is not changed.
For bitset: https://milvus.io/blog/2022-2-14-bitset.md

**Reply by failable** [reply]
Thank you!

**Reply by patelprateek** [reply]
@yhmo : to clarify , we still do this brute force predicate evaluation (parallelized across segments) and then do filtering .
What happens when post filtering number of embedding vector are high , that would be pretty costly computationally and latency wise ? 
Do we not support any inverted or sorted index based on scalar predicates , i read we have partition_keys but that is only applicable to one column or multiple ?

**Reply by yhmo** [reply]
You can create index for scalar fields.
`collection.create_index(field_name="xxx")`

For VARCHAR field, it creates Trie(prefix tree) index.
For numeric fields, it sorts the values as index.

**Comment by VijayDeyTR** [comment]
@yhmo Is is possible to add an index on the scaler field being specified in the hybrid search filter - this may reduce the search time?

https://milvus.io/docs/build_scalar_index.md

**Reply by yhmo** [reply]
Yeah, creating an index on a scalar field can improve hybrid search performance slightly.
In v2.2.9, we will introduce a new feature "partition key" to improve hybrid search significantly.

**Reply by sandszhouSZ** [reply]
Is there something new to share about this topic?

**Reply by yhmo** [reply]
> Is there something new to share about this topic?

Do you mean the feature "partition key"? It is already released in v2.2.9
